### DISCUSSION TITLE
Use of "terminal" versus "command line"

### DISCUSSION THREAD

**Discussion by julian-cable** [StyleGuides/WritingStyleGuide#613] [General]
In the style guide, the existing word usage entry for "terminal" states:

_Do not use to describe where to type commands. Use "command line" instead._ 

However, this guidance is not sufficiently clear (it has caused confusion for writers), and is possibly too restrictive as it stands. The guidance should cover when to use "terminal" as well as when not to use it. 

It seems to me that a distinction can be made between the use of "terminal" to refer to the session where certain activities are done, versus "command line" to refer to the specific prompt where a command is run. 

Some examples: 
"Open a terminal window and log in to your OpenShift cluster as the `admin` user with `redhatocp` as the password."
"The exercise file captures any error messages from your terminal."

"Terminate `process101`, `process102`, and `process103` from the command line."
"You can create custom storage pools from the command line."
"Use the command line to retrieve the route for the S3 endpoint and the bucket name."
"You can access the logs from the OpenShift Cluster Manager web interface, or from the command line by using the `rosa logs
install` command."

A related consideration is the possible use of "command-line interface" or "CLI" as alternative wording.

**Comment by gjbianco** [comment]
An additional point of confusion/contention is that not every terminal-based interface involves issuing commands. Some circles distinguish between a "CLI" and a "TUI" ("terminal user interface").

For example, Git is (usually) a CLI in that you enter commands that being with `git`, but things like terminal-based text editors would qualify as a TUI.

I'm not sure we want to adopt the TUI term, but there _are_ some terminal-based applications that arguably are not a "command-line interface" in that they don't accept commands (and/or that is not their primary form of interaction).

All of that being said, I think _most_ people use "CLI" and "terminal" rather interchangeably. The distinction _usually_ doesn't matter (same with CLI vs TUI). It's unfortunately something that might be really difficult for us to adopt a specific nomenclature around.

**Reply by bonnevil** [reply]
Usually, I'd distinguish between things like **nslookup** and **mysql** and even **fdisk** that have some form of command line that is used to enter commands, and something that provides a _curses_-based TUI like an old DOS program.  There's darn few of those left in RHEL, thank goodness.  Usually we'd talk about a TUI as something that would display an interface in a text-based terminal window like this:

![image](https://github.com/user-attachments/assets/48eee888-3b49-4dc9-8aa1-3026f17f74e3)

**Reply by bonnevil** [reply]
In fact, historically some of the tools that used _ncurses_ for this sort of interface had binary names that ended in **-tui** like **podman-tui**.   But there are also _ncurses_-based applications that have these "TUI" interfaces but do not follow that naming convention, like **iptraf-ng**.

**Comment by bonnevil** [comment]
Also, we have a bunch of very specific related terms that might come up.  In RH124, chapter 2 "Access the Command Line", the first and third presentations are relevant.  But I'll snip this from RH124 ch02s01, with some minor edits:

- A _command line_ is a text-based interface that provides a prompt to input typed instructions into a computer system.
- The Linux command line is provided by a program called the _shell_. Many shell program variants have been developed over the years. Every user can use a different shell, but Red Hat recommends using the default shell for system administration. The default user shell in Red Hat Enterprise Linux (RHEL) is the GNU Bourne-Again Shell (**bash**). The **bash** shell is an improved version of the original Bourne Shell (**sh**) on Unix systems.
- A _terminal_ is a text-based interface to enter commands into and print output from a computer system. To run the shell, you must log in to the computer on a terminal.
- The _physical console_ is a hardware keyboard and display, used for input and output, connected directly to the computer.
- The physical console supports multiple _virtual consoles_, which can each operate like a separate terminal. Each virtual console supports an independent login session. You can switch between the virtual consoles by pressing **Ctrl**+**Alt** and a function key (**F1** through **F6**) at the same time. Most of these virtual consoles run a terminal that provides a text login prompt. If you enter your username and password correctly, then you log in and get a shell prompt.
- The computer might provide a graphical login prompt on one of the virtual consoles. You can use the graphical login prompt to log in to a _graphical environment_. The graphical environment also runs on a virtual console. To get a shell prompt, you must start a terminal program in the graphical environment.
- A shell started by a graphical terminal program or remote SSH session runs on a _pseudoterminal_ that is not associated with a console.
- A _headless server_ does not have a keyboard and display that are permanently connected to it.  A data center might be filled with many racks of headless servers, and not providing each with a keyboard and display saves space and expense. For administrators to log in, a login prompt for a headless server might be provided by its _serial console_, which runs on a serial port that is connected to a networked console server for remote access.

This doesn't go into how programs like **getty** and **mingetty** work to provide a login prompt on the consoles.  

There's also a fine distinction between "command line" (anything that provides a prompt for text commands) and "shell" (the one used to interact with the Linux operating system's user space).  For example, you could argue that the MySQL client that you can start from the shell (**mysql**) has a command-line interface, but it's not a shell.

**Comment by julian-cable** [comment]
In the 7.0 release of the style guide, guidance was updated for the word usage entries: "command line", "command-line", "command prompt", "shell prompt", and "terminal".
